build_tools/therock_build_observability.py

```diff
@@ -0,0 +1,390 @@
+- NEVER fails the build if profiling/reporting fails.
+  
+- Only propagates the real compiler's exit code.
+
+Usage in CMake configure:
+
+  cmake -S . -B build -GNinja \
+    -DTHEROCK_AMDGPU_FAMILIES=gfx110X-all \
+    -DCMAKE_C_COMPILER_LAUNCHER="${PWD}/build_tools/therock_build_observability.py" \
+    -DCMAKE_CXX_COMPILER_LAUNCHER="${PWD}/build_tools/therock_build_observability.py"
+
+
+  cmake --build build 
+          OR 
+  ninja -C build -j"$(nproc)"
+
+
+Output: 
+    End summary for observability of build resources gets created in /tmp/comp-summary.md /tmp/comp-summary.csv file 
+
+
+"""
+
+import os
+import sys
+import time
+import random
+import datetime
+import resource
+import subprocess
+import shlex
+from typing import Dict, Tuple
+
+
+# -------------------------
+# Component classification
+# -------------------------
+
+def guess_component_from_pwd_and_cmd(pwd: str, cmd_str: str) -> str:
+    """ when not clear to infer which component this compile belongs to."""
+    comp = "unknown"
+
+    # Infer from build directory (PWD). Can be edited, added as needed
+    if "/core/" in pwd:
+        comp = "core"
+    elif "/compiler/" in pwd:
+        comp = "compiler"
+    elif "/math-libs/" in pwd:
+        comp = "math-libs"
+    elif "/ml-libs/" in pwd:
+        comp = "ml-libs"
+    elif "/profiler/" in pwd:
+        comp = "profiler"
+    elif "/dctools/" in pwd:
+        comp = "dctools"
+    elif "/rocm-libraries/" in pwd:
+        comp = "rocm-libraries"
+    elif "/rocm-systems/" in pwd:
+        comp = "rocm-systems"
+
+    # Define command text (per-library)
+    lower_cmd = cmd_str.lower()
+    if "rocblas" in lower_cmd:
+        comp = "rocblas"
+    elif "rocsolver" in lower_cmd:
+        comp = "rocsolver"
+    elif "rocfft" in lower_cmd:
+        comp = "rocfft"
+    elif "miopen" in lower_cmd:
+        comp = "miopen"
+    elif "rccl" in lower_cmd:
+        comp = "rccl"
+    elif "hipblaslt" in lower_cmd:
+        comp = "hipblaslt"
+
+    return comp
+
+
+# -------------------------
+# Logging + measurement
+# -------------------------
+
+def run_and_log_command(log_dir: str) -> int:
+    """
+    Run the actual compiler command (sys.argv[1:]),
+    log its timing & resources, and return the compiler's exit code.
+    """
+    if len(sys.argv) <= 1:
+        # Nothing to run – act as a no-op and don't fail the build.
+        return 0
+
+    os.makedirs(log_dir, exist_ok=True)
+
+    pwd = os.getcwd()
+    cmd_args = sys.argv[1:]
+    cmd_str = " ".join(shlex.quote(arg) for arg in cmd_args)
+
+    comp = guess_component_from_pwd_and_cmd(pwd, cmd_str)
+
+    # Unique log filename. note that the log files are not getting saved anywhere else other than /tmp and can be cleared post reporting operation 
+    ts = datetime.datetime.now().strftime("%Y%m%d-%H%M%S")
+    rand = random.randint(0, 999999)
+    log_file = os.path.join(log_dir, f"build-{ts}-{rand}-{comp}.log")
```

Comment:
Use `pathlib` instead of `os.path.join`: https://github.com/ROCm/TheRock/blob/main/docs/development/style_guide.md#use-pathlib-for-filesystem-paths